File: identity-android/src/androidTest/java/com/android/identity/android/direct_access/ShadowIsoDep.java
/*
 * Copyright 2025 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.identity.android.direct_access;

import android.nfc.Tag;
import com.android.identity.android.mdoc.deviceretrieval.IsoDepWrapper;
import com.android.identity.direct_access.DirectAccessTransport;

import java.io.IOException;

public class ShadowIsoDep implements IsoDepWrapper {

  public static String TECH_ISO_DEP = "android.nfc.tech.IsoDep";

  private DirectAccessTransport mTransport;
  private int mTimeout;

  public ShadowIsoDep(DirectAccessTransport transport) {
    mTransport = transport;
  }

  @Override
  public Tag getTag() {
    //android.nfc.tech.IsoDep
    //
    return null;
  }

  @Override
  public void getIsoDep(Tag tag) {
  }

  @Override
  public boolean isConnected() {
    try {
      return mTransport.isConnected();
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
  }

  @Override
  public void connect() throws IOException {
    mTransport.openConnection();
  }

  @Override
  public void close() throws IOException {
    mTransport.closeConnection();
  }

  @Override
  public boolean isTagSupported() {
    return true;
  }

  @Override
  public void setTimeout(int timeout) {
    mTimeout = timeout;
  }

  @Override
  public int getTimeout() {
    return mTimeout;
  }

  @Override
  public byte[] getHistoricalBytes() {
    return new byte[0];
  }

  @Override
  public byte[] getHiLayerResponse() {
    return new byte[0];
  }

  @Override
  public int getMaxTransceiveLength() {
    return mTransport.getMaxTransceiveLength();
  }

  @Override
  public byte[] transceive(byte[] data) throws IOException {
    return mTransport.sendData(data);
  }

  @Override
  public boolean isExtendedLengthApduSupported() {
    return true;
  }
}

File: wallet/src/androidTest/java/org/multipaz/android/direct_access/ShadowIsoDep.kt
package org.multipaz.android.direct_access

import android.nfc.Tag
import com.android.identity.android.mdoc.deviceretrieval.IsoDepWrapper
import java.io.IOException


class ShadowIsoDep : IsoDepWrapper {
    val TECH_ISO_DEP: String = "android.nfc.tech.IsoDep"

    private var mTimeout = 0

    override fun getTag(): Tag? {
        //android.nfc.tech.IsoDep
        //
        return null
    }

    override fun getIsoDep(tag: Tag?) {
    }

    override val isConnected: Boolean
        get() {
            try {
                return DirectAccessSmartCardTransport.isConnected
            } catch (e: IOException) {
                throw RuntimeException(e)
            }
        }

    @Throws(IOException::class)
    override fun connect() {
        DirectAccessSmartCardTransport.openConnection()
        DirectAccessSmartCardTransport.openConnection()
    }

    @Throws(IOException::class)
    override fun close() {
        DirectAccessSmartCardTransport.closeConnection()
    }

    override val isTagSupported: Boolean
        get() {
            return true
        }

    override fun setTimeout(timeout: Int) {
        mTimeout = timeout
    }

    override fun getTimeout(): Int {
        return mTimeout
    }

    override fun getHistoricalBytes(): ByteArray {
        return ByteArray(0)
    }

    override fun getHiLayerResponse(): ByteArray {
        return ByteArray(0)
    }

    override val maxTransceiveLength: Int
        get() {
            return DirectAccessSmartCardTransport.maxTransceiveLength
        }

    @Throws(IOException::class)
    override fun transceive(data: ByteArray?): ByteArray {
        return data?.let { DirectAccessSmartCardTransport.sendData(it) }!!
    }

    override val isExtendedLengthApduSupported: Boolean
        get() {
            return true
        }
}